src/tls/openssl/tls.c

```diff
@@ -321,69 +321,174 @@ int tls_set_selfsigned(struct tls *tls, const char *cn)
 }
 
 
-int tls_set_selfsigned_rsa(struct tls *tls, const char *cn, size_t bits)
+static int tls_generate_cert(X509 **pcert, const char *cn)
 {
+	X509 *cert = NULL;
 	X509_NAME *subj = NULL;
+	int e = 0;
+
+	if (!pcert || !cn)
+		goto err;
+
+	cert = X509_new();
+	if (!cert)
+		goto err;
+
+	if (!X509_set_version(cert, 2))
+		goto err;
+
+	if (!ASN1_INTEGER_set(X509_get_serialNumber(cert), rand_u32()))
+		goto err;
+
+	subj = X509_NAME_new();
+	if (!subj)
+		goto err;
+
+	if (!X509_NAME_add_entry_by_txt(subj, "CN", MBSTRING_ASC,
+					(unsigned char *)cn,
+					(int)strlen(cn), -1, 0))
+		goto err;
+
+	if (!X509_set_issuer_name(cert, subj) ||
+	    !X509_set_subject_name(cert, subj))
+		goto err;
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+	if (!X509_gmtime_adj(X509_getm_notBefore(cert), -3600*24*365) ||
+	    !X509_gmtime_adj(X509_getm_notAfter(cert),   3600*24*365*10))
+		goto err;
+#else
+	if (!X509_gmtime_adj(X509_get_notBefore(cert), -3600*24*365) ||
+	    !X509_gmtime_adj(X509_get_notAfter(cert),   3600*24*365*10))
+		goto err;
+#endif
+
+	goto out;
+
+ err:
+	e = 1;
+
+ out:
+	if (e)
+		X509_free(cert);
+	else
+		*pcert = cert;
+
+	X509_NAME_free(subj);
+	return e;
+}
+
+
+/**
+ * Create a selfsigned X509 certificate using EC
+ *
+ * @param tls   TLS Contect
+ * @param cn    Common Name
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int tls_set_selfsigned_ec(struct tls *tls, const char *cn)
+{
+	EC_KEY *eckey = NULL;
 	EVP_PKEY *key = NULL;
 	X509 *cert = NULL;
-	BIGNUM *bn = NULL;
-	RSA *rsa = NULL;
 	int r, err = ENOMEM;
 
 	if (!tls || !cn)
 		return EINVAL;
 
-	rsa = RSA_new();
-	if (!rsa)
+	eckey = EC_KEY_new_by_curve_name(NID_secp521r1);
```

Comment:
perhaps the curve name can be added as a function argument.
then the application can choose which one to use.